### ISSUE TITLE
RC1 Method returning String crashes godot

### ISSUE THREAD

**Issue by SpiceyWolf** [godot-dlang/godot-dlang#80]
Ive tried a good few combinations of things whether static or non static method, and different inputs in the params... seemingly it JUST seems to be Variant or String returns on @Method that are just crashing godot (And making the godot crash on bootup even when it hasnt had a chance to do a thing)

**Comment by Superbelko** [comment]
That's too vague, test project has string returns and it definitely works, more so godot won't call any methods by itself.

Try debugging this in vscode using following configuration, open your project and add a new debug launch configuration, but use these two instead (assuming code-d is installed), adjust paths, project and executable names where needed.

__.vscode/launch.json__:
```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(test project) GD4 editor",
            "type": "code-d",
            "request": "launch",
            "program": "/home/username/prog/godot/bin/godot.linuxbsd.editor.dev.x86_64",
            "args": ["${workspaceFolder}/examples/test/project/project.godot"],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "console": "internalConsole"
        },
        {
            "name": "(test project) GD4 instance",
            "type": "code-d",
            "request": "launch",
            "program": "/home/username/prog/godot/bin/godot.linuxbsd.editor.dev.x86_64",
            "args": ["--path", "${workspaceFolder}/examples/test/project"],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "console": "internalConsole"
        }
    ]
}
```

**Comment by Superbelko** [comment]
After fixing build error it looks like returning Variant is either not supported at all or it simply handled as Variant type derived from Object.

**Comment by SpiceyWolf** [comment]
This code has no issues:
```
module gd.test_cls;

import godot;
import godot.variant;
import godot.string;
import godot.refcounted;

class Testcls : GodotScript!RefCounted
{
    @Method
    static void to_hex(Variant input)
    {
        print(String(input.toString()));
    }
}
```

and this one crashes godot:
```
module gd.test_cls;

import godot;
import godot.variant;
import godot.string;
import godot.refcounted;

class Testcls : GodotScript!RefCounted
{
    @Method
    static String to_hex(Variant input)
    {
        return String(input.toString());
    }
}
```

Also i was looking at example code from the repo, and the string returns were with @Property tag.

-- As far as the debug file you listed above, how exactly do I begin using that once its entered?

**Comment by Superbelko** [comment]
On the side panel there a bug icon for debugging tab, inside this tab there a dropdown list on top to the left side where you can pick one of the configurations above, choose one and at any time hit F5 to start debugging.

Ok this example might help, likely it crashes in the second one because string is gone by the time godot receives it, or something like that, or there is something messed up inside Variant.toString()

`@Property` is for getter/setter/backing variable, very unlikely that it is related to this problem at all.

**Comment by SpiceyWolf** [comment]
Boutta clock in so cant test anything, but last night i ttied just returning a String("hi") and it failed, String as param input as well and just returning the same string back... Its def not the variant there causing it, was only when selecting variant or string as the method return type

**Comment by Superbelko** [comment]
Test project again has some issues with array.slice(),
but aside from that I did this and it works just fine

```d
@Method
String retString(String s) { return s; }

@Method
String retVarString(Variant v) { return String(v.toString()); }
```

```gdscript
extends Node

func _ready():
	var t: TestD = TestD.new()
	var v: Variant = "test string in variant"
	print(t.ret_string("test string to D side"))
	print(t.ret_var_string(v))
```

output as expected
```
test string to D side
test string in variant
```

**Comment by SpiceyWolf** [comment]
Ill be home in ~2 hours, will you be around then? And second to that, are you in the community discord?

**Comment by SpiceyWolf** [comment]
Might be linux exclusive of an issue? I had the idea "maybe" it was a static related issue and tried JUST a standalone string ret... this in fact crashes:

```
    @Method
    String retStr(String input)
    {
        return input;
    }
    ```

**Comment by Superbelko** [comment]
I've fixed methods that returns Variant (#81), previously it was unable to compile it.

---

Your retStr function works just fine, it could be some RC1 bug, I use custom debug build of godot master so that maybe why I don't have this.

---

DMD vs LDC has different optimizations and code generation capabilities, sometimes this happens and code that works fine with one of them fails with another one, can't help you there.

Sometimes it has just even stupid issues like this one, it used in type reflection, which probably has some issue in how LDC implements internal `__traits(compiles,...)` or standard library functions:
https://github.com/godot-dlang/godot-dlang/blob/88d4cff1578613df5773f6f844c247e353edaf28/src/godot/api/wrap.d#L616

---

> No symbol "GDExtensionClassMethodArgumentMetadata" in current context

You use godot release build without debug info, that what this message likely means.

If you really want to, you can have debug build, just get `python3` and `scons`, clone the godot repo and from godot folder just run

```scons platform=x11 dev_build=yes```

on my machine it is just 3 minutes.

**Comment by SpiceyWolf** [comment]
K ill snag tonight
